src/validators/recursive.rs

```diff
@@ -77,19 +77,24 @@ fn validate<'s, 'data>(
     input: &'data impl Input<'data>,
     extra: &Extra,
     slots: &'data [CombinedValidator],
-    recursion_guard: &'s mut RecursionGuard,
+    recursion_guard: &'s mut Option<&mut RecursionGuard>,
 ) -> ValResult<'data, PyObject> {
+    let new_guard = &mut RecursionGuard::new();
+    let guard = match recursion_guard {
+        Some(guard) => guard,
+        None => new_guard,
+    };
```

Comment:
Couldn't get the borrow checker to be okay with allocating inside the match scope, which forces the allocation every time even if we already had a RecursionGuard. So unless this can be fixed recursive validation would be slower, but everything else should see near zero performance impact because of this check (just passing around an extra reference to an option).